workflow/archive/archive_controller.go

```diff
@@ -0,0 +1,187 @@
+package archive
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	syncpkg "github.com/argoproj/pkg/sync"
+	log "github.com/sirupsen/logrus"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	runtimeutil "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/argoproj/argo-workflows/v3/persist/sqldb"
+	wfclientset "github.com/argoproj/argo-workflows/v3/pkg/client/clientset/versioned"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+	"github.com/argoproj/argo-workflows/v3/workflow/hydrator"
+	"github.com/argoproj/argo-workflows/v3/workflow/metrics"
+	"github.com/argoproj/argo-workflows/v3/workflow/util"
+)
+
+type Controller struct {
+	wfclientset     wfclientset.Interface
+	wfInformer      cache.SharedIndexInformer
+	wfArchiveQueue  workqueue.RateLimitingInterface
+	hydrator        hydrator.Interface
+	wfArchive       sqldb.WorkflowArchive
+	workflowKeyLock syncpkg.KeyLock
```

Comment:
This is no longer the same lock as the one used in the main controller. It feels like this might matter, does it?

Reply:
Yes, I think pass the lock by pointer can fix it

Reply:
This isn't a pointer, it needs to be for this to work.